File: frontend/src/lib/api/plants.ts
import { deleteResource, getJson, postJson, putJson } from "./client";

/**
 * Pot material wire values (design §1 / D3). String-literal union matching the
 * backend `PotMaterial` StrEnum members exactly.
 */
export type PotMaterial =
  | "terracotta"
  | "plastic"
  | "ceramic"
  | "self-watering"
  | "other";

/**
 * Light-level wire values (design §1 / D3). String-literal union matching the
 * backend `LightLevel` StrEnum members exactly.
 */
export type LightLevel = "dark" | "indirect" | "bright-indirect" | "full-sun";

/** All `PotMaterial` values, for rendering the select options. */
export const POT_MATERIALS: readonly PotMaterial[] = [
  "terracotta",
  "plastic",
  "ceramic",
  "self-watering",
  "other",
];

/** All `LightLevel` values, for rendering the select options. */
export const LIGHT_LEVELS: readonly LightLevel[] = [
  "dark",
  "indirect",
  "bright-indirect",
  "full-sun",
];

/**
 * One enabled schedule's due state as returned inside `PlantResponse.schedules`
 * (the additive US-3.3 field; `ScheduleDueResponse` on the wire). `care_type` is
 * the closed schedule vocabulary `water`/`feed` - NOT the four-member
 * `CareEventType`. `next_due` is null only when the schedule is paused/dormant
 * inside the window; `overdue_days` is null iff `next_due` is null (the
 * both-null invariant), else `>= 0`.
 */
export interface ScheduleDue {
  readonly care_type: "water" | "feed";
  readonly next_due: string | null;
  readonly overdue_days: number | null;
}

/**
 * A persisted plant as returned by `GET /api/v1/plants` (`PlantResponse`).
 *
 * Server response fields are snake_case (design §1). `location_id` is null when
 * the plant is homeless (D-009). No runtime validation is performed; the shape
 * mirrors `PlantResponse`.
 */
export interface Plant {
  readonly id: number;
  readonly name: string;
  readonly species: string | null;
  readonly location_id: number | null;
  readonly acquired_on: string | null;
  readonly pot_size_cm: number | null;
  readonly pot_material: PotMaterial | null;
  readonly light_level: LightLevel | null;
  readonly notes: string | null;
  readonly tags: readonly string[];
  readonly archived: boolean;
  readonly cover_photo_id: number | null;
  /**
   * One entry per enabled schedule of the plant, carrying its computed due
   * state (US-3.3, additive; already on the wire). Archived plants and plants
   * with no enabled schedule carry `[]`.
   */
  readonly schedules: readonly ScheduleDue[];
  readonly created_at: string;
  readonly updated_at: string;
}

/** Request body for creating/updating a plant (`PlantCreate`/`PlantUpdate`). */
export interface PlantInput {
  readonly name: string;
  readonly species: string | null;
  readonly location_id: number | null;
  readonly acquired_on: string | null;
  readonly pot_size_cm: number | null;
  readonly pot_material: PotMaterial | null;
  readonly light_level: LightLevel | null;
  readonly notes: string | null;
  readonly tags: readonly string[];
  readonly archived: boolean;
}

/**
 * Optional search/filter params for the list endpoint (design §1 / D4).
 * All optional and AND-combined server-side. `homeless=true` returns only
 * null-location plants. Unset / empty fields are omitted from the query string.
 */
export interface PlantFilter {
  readonly q?: string;
  readonly location_id?: number | null;
  readonly tag?: string;
  readonly species?: string;
  readonly homeless?: boolean;
  readonly archived?: boolean;
  readonly include_archived?: boolean;
}

/**
 * Render a `PlantFilter` to a `?...` query string. Only set, non-empty fields
 * are appended (D4 - all params optional); an empty/undefined filter yields the
 * bare collection path. `homeless` is rendered only when true.
 */
function buildQuery(filter: PlantFilter | undefined): string {
  if (filter === undefined) {
    return "";
  }
  const params = new URLSearchParams();
  if (filter.q !== undefined && filter.q.length > 0) {
    params.set("q", filter.q);
  }
  if (filter.location_id !== undefined && filter.location_id !== null) {
    params.set("location_id", String(filter.location_id));
  }
  if (filter.tag !== undefined && filter.tag.length > 0) {
    params.set("tag", filter.tag);
  }
  if (filter.species !== undefined && filter.species.length > 0) {
    params.set("species", filter.species);
  }
  if (filter.homeless === true) {
    params.set("homeless", "true");
  }
  // `archived` is a tri-state scope (A2): unset -> active (server default),
  // true -> archived only, false -> active only. Render whenever set.
  if (filter.archived !== undefined) {
    params.set("archived", String(filter.archived));
  }
  // `include_archived` overrides the scope to "all"; render only when true
  // (mirrors `homeless`).
  if (filter.include_archived === true) {
    params.set("include_archived", "true");
  }
  const query = params.toString();
  return query.length > 0 ? `?${query}` : "";
}

/**
 * `GET /api/v1/plants` - all plants ordered by name, narrowed by the optional
 * filter. Throws `ApiError` on non-2xx.
 */
export function fetchPlants(filter?: PlantFilter): Promise<Plant[]> {
  return getJson<Plant[]>(`/plants${buildQuery(filter)}`);
}

/** `GET /api/v1/plants/{id}` - one plant. Throws `ApiError` on non-2xx. */
export function fetchPlant(id: number): Promise<Plant> {
  return getJson<Plant>(`/plants/${String(id)}`);
}

/** `POST /api/v1/plants` - create a plant (201). Throws `ApiError` on non-2xx. */
export function createPlant(input: PlantInput): Promise<Plant> {
  return postJson<Plant>("/plants", input);
}

/** `PUT /api/v1/plants/{id}` - full-replace update (200). Throws `ApiError` on non-2xx. */
export function updatePlant(id: number, input: PlantInput): Promise<Plant> {
  return putJson<Plant>(`/plants/${String(id)}`, input);
}

/** `DELETE /api/v1/plants/{id}` - remove a plant (204). Throws `ApiError` on non-2xx. */
export function deletePlant(id: number): Promise<void> {
  return deleteResource(`/plants/${String(id)}`);
}

/**
 * `POST /api/v1/plants/{id}/archive` - idempotent state-set action (US-2.4 / A1).
 * Empty body; returns the updated `PlantResponse` (200). Throws `ApiError` on
 * non-2xx (incl. 404 for an unknown id).
 */
export function archivePlant(id: number): Promise<Plant> {
  return postJson<Plant>(`/plants/${String(id)}/archive`, {});
}

/**
 * `POST /api/v1/plants/{id}/unarchive` - idempotent state-set action (US-2.4 / A1).
 * Empty body; returns the updated `PlantResponse` (200). Throws `ApiError` on
 * non-2xx (incl. 404 for an unknown id).
 */
export function unarchivePlant(id: number): Promise<Plant> {
  return postJson<Plant>(`/plants/${String(id)}/unarchive`, {});
}

File: backend/src/viridarium/adapters/inbound/web/plants.py
"""Plants router (ARCH-002: HTTP only, no business logic).

Maps the wire schemas to/from domain types and delegates to the
:class:`~viridarium.application.plants.PlantService`. The list endpoint reads the
optional query params into a :class:`~viridarium.domain.plant.PlantFilter`. Domain
errors (``PlantNotFoundError`` -> 404, ``LocationNotFoundForPlantError`` -> 422) are
not caught here: they are translated by the registered exception handlers in the app
factory (ADR-C), keeping status codes out of the router.
"""

from __future__ import annotations

from typing import Annotated

from fastapi import APIRouter, Depends, Query, Response, status

from viridarium.adapters.inbound.web.dependencies import (
    get_due_query_service,
    get_plant_service,
)
from viridarium.adapters.inbound.web.schemas import (
    PlantCreate,
    PlantResponse,
    PlantUpdate,
    ScheduleDueResponse,
)
from viridarium.application.due import DueQueryService
from viridarium.application.plants import PlantService
from viridarium.domain.plant import NewPlant, Plant, PlantFilter

router = APIRouter(prefix="/plants", tags=["plants"])

ServiceDep = Annotated[PlantService, Depends(get_plant_service)]
DueServiceDep = Annotated[DueQueryService, Depends(get_due_query_service)]


def _with_due(plant: Plant, due: DueQueryService) -> PlantResponse:
    """Compose a plant read with its computed ``schedules`` due field (US-3.3).

    An archived plant is excluded from due computation entirely (empty schedules); a
    non-archived plant gets one entry per enabled schedule. The field is composed here,
    not read off the domain ``Plant`` (which has no due).
    """
    response = PlantResponse.model_validate(plant)
    if plant.archived:
        return response
    due_by_plant = due.for_plants([plant.id])
    response.schedules = [
        ScheduleDueResponse.from_domain(d) for d in due_by_plant.get(plant.id, [])
    ]
    return response


def _to_new_plant(body: PlantCreate) -> NewPlant:
    return NewPlant(
        name=body.name,
        species=body.species,
        location_id=body.location_id,
        acquired_on=body.acquired_on,
        pot_size_cm=body.pot_size_cm,
        pot_material=body.pot_material,
        light_level=body.light_level,
        notes=body.notes,
        tags=tuple(body.tags),
        archived=body.archived,
    )


@router.post(
    "",
    response_model=PlantResponse,
    status_code=status.HTTP_201_CREATED,
    summary="Create a plant",
)
def create_plant(body: PlantCreate, service: ServiceDep) -> PlantResponse:
    """Create a new plant."""
    created = service.create(_to_new_plant(body))
    return PlantResponse.model_validate(created)


@router.get("", response_model=list[PlantResponse], summary="List plants")
def list_plants(
    service: ServiceDep,
    due: DueServiceDep,
    q: Annotated[str | None, Query()] = None,
    location_id: Annotated[int | None, Query()] = None,
    tag: Annotated[str | None, Query()] = None,
    species: Annotated[str | None, Query()] = None,
    homeless: Annotated[bool, Query()] = False,
    archived: Annotated[bool | None, Query()] = None,
    include_archived: Annotated[bool, Query()] = False,
) -> list[PlantResponse]:
    """List plants, optionally filtered (AND-combined), ordered by name.

    Defaults to active plants only; ``archived=true`` returns archived only and
    ``include_archived=true`` returns all (US-2.4).
    """
    plant_filter = PlantFilter(
        q=q,
        location_id=location_id,
        tag=tag,
        species=species,
        homeless=homeless,
        archived=archived,
        include_archived=include_archived,
    )
    plants = service.list(plant_filter)
    # One batch due read for all non-archived ids -> the list path stays flat (no N+1,
    # AC7); archived plants are excluded from the computation and get empty schedules.
    active_ids = [p.id for p in plants if not p.archived]
    due_by_plant = due.for_plants(active_ids)
    responses: list[PlantResponse] = []
    for plant in plants:
        response = PlantResponse.model_validate(plant)
        response.schedules = [
            ScheduleDueResponse.from_domain(d) for d in due_by_plant.get(plant.id, [])
        ]
        responses.append(response)
    return responses


@router.get("/{plant_id}", response_model=PlantResponse, summary="Get a plant")
def get_plant(plant_id: int, service: ServiceDep, due: DueServiceDep) -> PlantResponse:
    """Get one plant by id, with its computed ``schedules`` due field (US-3.3)."""
    return _with_due(service.get(plant_id), due)


@router.put("/{plant_id}", response_model=PlantResponse, summary="Replace a plant")
def update_plant(
    plant_id: int, body: PlantUpdate, service: ServiceDep
) -> PlantResponse:
    """Full-replace a plant's fields and tags (ADR-D)."""
    updated = service.update(plant_id, _to_new_plant(body))
    return PlantResponse.model_validate(updated)


@router.delete(
    "/{plant_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="Delete a plant",
)
def delete_plant(plant_id: int, service: ServiceDep) -> Response:
    """Delete a plant by id."""
    service.delete(plant_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.post(
    "/{plant_id}/archive",
    response_model=PlantResponse,
    summary="Archive a plant",
)
def archive_plant(plant_id: int, service: ServiceDep) -> PlantResponse:
    """Archive a plant (idempotent state-set); 404 on an unknown id (A1)."""
    return PlantResponse.model_validate(service.archive(plant_id))


@router.post(
    "/{plant_id}/unarchive",
    response_model=PlantResponse,
    summary="Unarchive a plant",
)
def unarchive_plant(plant_id: int, service: ServiceDep) -> PlantResponse:
    """Unarchive a plant (idempotent state-set); 404 on an unknown id (A1)."""
    return PlantResponse.model_validate(service.unarchive(plant_id))
